docs/core/diagnostics/eventsource.md

```diff
@@ -20,6 +20,17 @@ your own custom events.
 > Many technologies that integrate with EventSource use the terms 'Tracing' and 'Traces' instead of 'Logging' and 'Logs'.
 > The meaning is the same here.
 
+> [!WARNING]
+> Potential Deadlocks with EventSource Callbacks
+>
+> When implementing EventSources, be cautious with invoking lock-acquiring APIs within EventSource callbacks. EventSource instances are initialized early in the runtime, before some core features are fully initialized. As a result, the re-entrant behavior of EventSource callbacks may cause deadlocks if the callback attempts to acquire locks already held by the thread that triggered the callback.
```

Comment:
Most locks in the runtime are re-entrant so it usually isn't a problem to acquire a lock the callback thread already holds. The most typical deadlock situation is the caller acquires one lock and the callback attempts to acquire a 2nd one that is not already held. Meanwhile some other thread has already acquired that 2nd lock and is blocked trying to acquire the 1st.

I don't think we need to get into a full description though. It should be sufficient to just warn that acquiring locks inside a callback where the thread normally would not have done so can be a cause of deadlocks.

Reply:
That makes sense, thank you for clearing up my misunderstanding!